doc/adr/2019-02-13-preserve-all-functionality.md

```diff
@@ -0,0 +1,31 @@
+**Title**: Preserve all functionality offered by the raw API.
+
+**Status**: Proposed
+
+**Context**: When we design APIs for the client libraries one may need to
+tradeoff ease of use for functionality. All things being equal it is easier to
+design simpler, easier to understand client libraries if one is willing to
+remove functionality that is rarely used. While we want to encourage users to
+adopt the libraries, we do not want to abandon advanced users to their own
+devices. We do not want them to be tempted or forced to discard the client
+libraries in favor of raw REST or gRPC because that is the only way they can
+access a feature.
+
+**Decision**: The libraries must preserve all functionality. If it is possible
+to perform a task using the raw gRPC or REST API, it should be possible to
```

Comment:
Note: This may affect our tentative plans around upcoming things like returning full spanner rows rather than streaming them in chunks like the gRPC API might do. We could certainly provide both options, but just pointing this out.

Reply:
Agree.  The language in this ADR is maximal, but in practice we have more subtle tradeoffs.  What do you think of something like:

```
The libraries should preserve all functionality. If it is possible to perform a task using the raw gRPC or REST API, it should be possible to use the client libraries to perform the same task.

Any time when we need to deviate from this principle, for example to provide a simpler API, we should document such deviations as a bug, and highlight this problem in the documentation.
```

Realistically we are always trading off simplicity for full functionality, if nothing else we consume a sliver of CPU to make the APIs simpler, and that stops users from doing N ops/second by a tiny-wee-bit of a percentage (it may be so small that we cannot measure it, but we know it is there).

Let me take a stab to writing more nuanced language.